service_discovery/mci_utils/mci.go

```diff
@@ -0,0 +1,316 @@
+	}
+
+	cl := make(map[string]interface{})
+	for _, cname := range clusterList {
+		cl[cname] = struct{}{}
+	}
+
+	// for each config:
+	//    - cluster context can't be empty
+	//    - cluster context must be present in clusterset
+	//    - at least one service must be there in the backend config
+	//    - service name can't be empty
+	for _, config := range mciObj.Spec.Config {
+		if config.ClusterContext == "" {
+			return fmt.Errorf("spec.config.clusterContext can't be empty")
+		}
+		_, clusterContextPresent := cl[config.ClusterContext]
+		if !clusterContextPresent {
+			return fmt.Errorf("cluster context %s is invalid and not part of clusterset", config.ClusterContext)
+		}
+		s := config.Service
+		if s.Name == "" {
+			return fmt.Errorf("service name can't be empty")
+		}
+		if s.Namespace == "" {
+			return fmt.Errorf("service namespace can't be empty")
+		}
+	}
+
+	return nil
+}
+
+type MCIServiceElement struct {
+	cluster   string
+	namespace string
+	name      string
+}
+
+func InitServiceElement(cname, namespace, name string) *MCIServiceElement {
+	return &MCIServiceElement{
+		cluster:   cname,
+		namespace: namespace,
+		name:      name,
+	}
+}
+
+func (se *MCIServiceElement) Cluster() string {
+	return se.cluster
+}
+
+func (se *MCIServiceElement) Namespace() string {
+	return se.namespace
+}
+
+func (se *MCIServiceElement) Name() string {
+	return se.name
+}
+
+func GetServiceList(mciObj *mciapi.MultiClusterIngress) ([]*MCIServiceElement, error) {
+	if mciObj == nil {
+		return nil, fmt.Errorf("error in getting service list as MCI object is nil")
+	}
+
+	mciSvcList := []*MCIServiceElement{}
+	for _, c := range mciObj.Spec.Config {
+		mciSvcList = append(mciSvcList, InitServiceElement(c.ClusterContext,
+			c.Service.Namespace, c.Service.Name))
+	}
+	return mciSvcList, nil
+}
+
+func GetServiceListStr(mciObj *mciapi.MultiClusterIngress) []string {
+	svcList := []string{}
+	for _, c := range mciObj.Spec.Config {
+		svcList = append(svcList, c.ClusterContext, c.Service.Namespace, c.Service.Name)
+	}
+	return svcList
+}
+
+func GetMCIServicesChecksum(mciObj *mciapi.MultiClusterIngress) uint32 {
+	svcList := GetServiceListStr(mciObj)
+	sort.Strings(svcList)
+	return containerutils.Hash(containerutils.Stringify(svcList))
+}
+
+func GetMCIsDiff(oldMCI, newMCI *mciapi.MultiClusterIngress) ([]*MCIServiceElement, []*MCIServiceElement, error) {
+	svcToAdd := []*MCIServiceElement{}
+	svcToDelete := []*MCIServiceElement{}
+
+	oldSvcs, err := GetServiceList(oldMCI)
+	if err != nil {
+		return svcToAdd, svcToDelete, fmt.Errorf("error in getting svc list from old MCI: %v", err)
+	}
+	newSvcs, err := GetServiceList(newMCI)
+	if err != nil {
+		return svcToAdd, svcToDelete, fmt.Errorf("error in getting svc list from new MCI: %v", err)
+	}
+
+	// generate a map of both old and new services
+	oldSvcsMap := make(map[string]interface{})
+	for _, s := range oldSvcs {
+		oldSvcsMap[s.Cluster()+s.Namespace()+s.Name()] = struct{}{}
```

Comment:
Would it be better to add a string like "/" between cluster, namespace and name ? Else foo in namesapce bar would get mapped to same key as rfoo in namespace fo

Reply:
Yeah, there's a problem in the logic. Thanks! Fixed now.